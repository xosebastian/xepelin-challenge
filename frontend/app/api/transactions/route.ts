import axios, { HttpStatusCode, isAxiosError } from "axios";
import { NextResponse } from "next/server";

const API_URL = process.env.API_URL;
const TRANSACTIONS_ENDPOINT = "transactions";

export async function POST(request: Request) {
  const { accountId, balance, type } = await request.json();

  if (!API_URL) {
    return NextResponse.json(
      { message: "API_URL is not defined" },
      {
        status: HttpStatusCode.InternalServerError,
      }
    );
  }

  try {
    const response = await axios.post(`${API_URL}/${TRANSACTIONS_ENDPOINT}`, {
      accountId,
      amount: balance,
      type,
    });

    return NextResponse.json(response.data, {
      status: HttpStatusCode.Created,
    });
  } catch (error) {
    if (isAxiosError(error)) {
      return NextResponse.json(error.response?.data, {
        status: error.response?.status ?? HttpStatusCode.InternalServerError,
      });
    }

    return NextResponse.json(
      { message: "Something went wrong" },
      {
        status: HttpStatusCode.InternalServerError,
      }
    );
  }
}
