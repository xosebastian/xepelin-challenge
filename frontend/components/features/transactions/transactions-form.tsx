"use client";

import { zodResolver } from "@hookform/resolvers/zod";
import * as z from "zod";

import { Button } from "@/components/ui/button";
import {
  Form,
  FormControl,
  FormField,
  FormItem,
  FormLabel,
  FormMessage,
} from "@/components/ui/form";
import { Input } from "@/components/ui/input";
import { useForm } from "react-hook-form";
import {
  Select,
  SelectContent,
  SelectItem,
  SelectTrigger,
} from "@/components/ui/select";
import { SelectValue } from "@radix-ui/react-select";

export const formTransactionSchema = z.object({
  balance: z
    .string()
    .max(4, {
      message: "The balance should have at most 4 digits.",
    })
    .regex(/^\d+(\.\d{1,2})?$/, {
      message: "The balance should be a number with up to 2 decimal places.",
    })
    .transform((value) => parseFloat(value)),
  type: z.string({
    required_error: "Please select an Type to display.",
  }),
});

interface Props {
  onSubmit: (values: z.infer<typeof formTransactionSchema>) => void;
  onChange?: () => void;
}

export function TransactionsForm({ onSubmit, onChange }: Props) {
  const form = useForm<z.infer<typeof formTransactionSchema>>({
    resolver: zodResolver(formTransactionSchema),
    defaultValues: {
      balance: 0,
      type: "DEPOSIT",
    },
  });

  return (
    <>
      <Form {...form}>
        <form onSubmit={form.handleSubmit(onSubmit)} onChange={onChange} className="space-y-2 pl-6">
          <FormField
            control={form.control}
            name="type"
            render={({ field }) => (
              <FormItem>
                <FormLabel>Operation Type</FormLabel>
                <Select
                  onValueChange={field.onChange}
                  defaultValue={field.value}
                  data-testid="type"
                >
                  <FormControl>
                    <SelectTrigger>
                      <SelectValue placeholder="Select the type of operation" />
                    </SelectTrigger>
                  </FormControl>
                  <SelectContent>
                    <SelectItem value="DEPOSIT">Deposit</SelectItem>
                    <SelectItem value="WITHDRAWAL">Withdraw</SelectItem>
                  </SelectContent>
                </Select>
                <FormMessage />
              </FormItem>
            )}
          />
          <FormField
            control={form.control}
            name="balance"
            render={({ field }) => (
              <FormItem>
                <FormLabel>Balance</FormLabel>
                <FormControl>
                  <Input data-testid="balance" type="number" {...field} />
                </FormControl>
                <FormMessage />
              </FormItem>
            )}
          />
          <Button role="button" type="submit">
            Execute
          </Button>
        </form>
      </Form>
    </>
  );
}
