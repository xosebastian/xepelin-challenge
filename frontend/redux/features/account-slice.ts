import { createSlice } from '@reduxjs/toolkit'
import type { PayloadAction } from '@reduxjs/toolkit'

export interface AccountState {
  id: string,
  balance: number,
}

const initialState: AccountState = {
  id: '',
  balance: 0,
}

export const accountSlice = createSlice({
  name: 'account',
  initialState,
  reducers: {
    create: (state, action: PayloadAction<{id : string}>) => {
      state.id = action.payload.id
    },
    clear: () => {
      return initialState
    },
    update: (state, action: PayloadAction<{balance: number}>) => {
      state.balance = action.payload.balance
    },
    transactions: (state, action: PayloadAction<{balance: number, type: string}>) => {
      if (action.payload.type === 'DEPOSIT') {
        state.balance += action.payload.balance
      }else if (action.payload.type === 'WITHDRAWAL') {
        state.balance -= action.payload.balance
      }
    },
  },
})


export const { create, clear, update, transactions } = accountSlice.actions

export default accountSlice.reducer