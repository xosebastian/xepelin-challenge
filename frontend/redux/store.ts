import { configureStore } from '@reduxjs/toolkit'
import accountReducer from './features/account-slice'
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux'

export const store = configureStore({
  reducer: {
    account: accountReducer,
  },
})

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;