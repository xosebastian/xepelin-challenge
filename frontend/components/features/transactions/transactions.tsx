"use client";
import { Alert, AlertDescription, AlertTitle } from "@/components/ui/alert";
import { useToast } from "@/components/ui/use-toast";
import { useAppDispatch, useAppSelector } from "@/redux/store";
import axios, { isAxiosError } from "axios";
import { useState } from "react";
import { z } from "zod";
import { TransactionsForm, formTransactionSchema } from "./transactions-form";
import { transactions } from "@/redux/features/account-slice";

export function Transactions() {
  const dispatch = useAppDispatch();
  const { toast } = useToast();
  const [error, setError] = useState<string | null>(null);
  const accountId = useAppSelector((state) => state.account.id);

  async function onSubmit(values: z.infer<typeof formTransactionSchema>) {
    const { balance, type } = values;

    try {
      await axios.post("/api/transactions", {
        accountId,
        balance,
        type,
      });

      dispatch(
        transactions({
          balance,
          type,
        })
      );
      toast({
        title: type,
        description: `${type} was successful`,
      });
    } catch (error: any) {
      const message = isAxiosError(error)
        ? error?.response?.data.message
        : error?.message ?? "Something went wrong";
      setError(message);
    }
  }

  return (
    <>
      {accountId ? (
        <>
          <div className="flex flex-col space-y-1.5 p-6">
            <h3 className="text-2xl font-semibold leading-none tracking-tight">
              Account Operations
            </h3>
          </div>
          <TransactionsForm
            onSubmit={onSubmit}
            onChange={() => setError(null)}
          />

          {error && (
            <Alert variant={"destructive"} className="mt-5">
              <AlertTitle>Ups!</AlertTitle>
              <AlertDescription>{error}</AlertDescription>
            </Alert>
          )}
        </>
      ) : (
        <div className="flex flex-col space-y-1.5 p-6">
          <h3 className="text-2xl font-semibold leading-none tracking-tight">
            Please create an account to continue
          </h3>
        </div>
      )}
    </>
  );
}
