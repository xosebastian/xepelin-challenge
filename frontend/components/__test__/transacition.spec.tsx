import { act, fireEvent, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import axios from "axios";
import { Transactions } from "../features/transactions/transactions";
import { Provider } from "react-redux";
import { store } from "@/redux/store";
import { create, clear } from "@/redux/features/account-slice";

jest.mock("axios");

const mockedAxios = axios as jest.Mocked<typeof axios>;
describe("Transactions component", () => {
  beforeEach(() => {
    mockedAxios.post.mockReset();
    store.dispatch(clear());
  });

  it("renders the account operations section", () => {
    store.dispatch(
      create({
        id: "user-id",
      })
    );

    const { container } = render(
      <Provider store={store}>
        <Transactions />
      </Provider>
    );
    expect(container).toMatchSnapshot();
    const accountOperations = screen.getByText("Account Operations");
    expect(accountOperations).toBeDefined();
  });

  it("shows a success toast when transaction is successful", async () => {
    mockedAxios.post.mockResolvedValueOnce({
      data: { balance: 100 },
    });

    store.dispatch(
      create({
        id: "user-id",
      })
    );

    render(
      <Provider store={store}>
        <Transactions />
      </Provider>
    );
    const balanceInput = screen.getByTestId("balance");
    const submitButton = screen.getByRole("button", { name: "Execute" });
    await act(async () => {
      await userEvent.type(balanceInput, "100");
      fireEvent.click(submitButton);
    });

    const balance = store.getState().account.balance;

    expect(balance).toEqual(100);
  });

  it("shows an error message when transaction fails", async () => {
    store.dispatch(
      create({
        id: "user-id",
      })
    );

    mockedAxios.post.mockRejectedValueOnce({
        message: "Insufficient funds",
    });
    render(
      <Provider store={store}>
        <Transactions />
      </Provider>
    );
    const balanceInput = screen.getByTestId("balance");
    const submitButton = screen.getByRole("button", { name: "Execute" });
    await act(async () => {
      await userEvent.type(balanceInput, "100");
      fireEvent.click(submitButton);
    });
    expect(screen.getByText("Insufficient funds")).toBeDefined();
  });

  it("shows a generic error message when transaction fails with no error message", async () => {
    store.dispatch(
      create({
        id: "user-id",
      })
    );
    mockedAxios.post.mockRejectedValueOnce({});
    render(
      <Provider store={store}>
        <Transactions />
      </Provider>
    );
    const balanceInput = screen.getByTestId("balance");
    const submitButton = screen.getByRole("button", { name: "Execute" });
    await act(async () => {
      await userEvent.type(balanceInput, "100");
      fireEvent.click(submitButton);
    });
    expect(screen.getByText("Something went wrong")).toBeDefined();
  });
});
