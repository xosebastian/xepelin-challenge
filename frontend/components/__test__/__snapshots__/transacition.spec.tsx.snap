// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transactions component renders the account operations section 1`] = `
<div>
  <div
    class="flex flex-col space-y-1.5 p-6"
  >
    <h3
      class="text-2xl font-semibold leading-none tracking-tight"
    >
      Account Operations
    </h3>
  </div>
  <form
    class="space-y-2 pl-6"
  >
    <div
      class="space-y-2"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for=":r0:-form-item"
      >
        Operation Type
      </label>
      <button
        aria-autocomplete="none"
        aria-controls="radix-:r1:"
        aria-describedby=":r0:-form-item-description"
        aria-expanded="false"
        aria-invalid="false"
        class="flex h-10 w-full items-center justify-between rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background placeholder:text-muted-foreground focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
        data-state="closed"
        dir="ltr"
        id=":r0:-form-item"
        role="combobox"
        type="button"
      >
        <span
          style="pointer-events: none;"
        >
          Deposit
        </span>
        <svg
          aria-hidden="true"
          class="h-4 w-4 opacity-50"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m6 9 6 6 6-6"
          />
        </svg>
      </button>
      <select
        aria-hidden="true"
        style="position: absolute; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
        tabindex="-1"
      >
        <option
          selected=""
          value="DEPOSIT"
        >
          Deposit
        </option>
        <option
          value="WITHDRAWAL"
        >
          Withdraw
        </option>
      </select>
    </div>
    <div
      class="space-y-2"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for=":r2:-form-item"
      >
        Balance
      </label>
      <input
        aria-describedby=":r2:-form-item-description"
        aria-invalid="false"
        class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
        data-testid="balance"
        id=":r2:-form-item"
        name="balance"
        type="number"
        value="0"
      />
    </div>
    <button
      class="inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
      role="button"
      type="submit"
    >
      Execute
    </button>
  </form>
</div>
`;
